### PR TITLE
Implement Node Id on Server

### LINKED ISSUE #1050: Generate a unique ID for each node
## Overview
**Feature request.**  Currently, the layout uses IP addresses to identify nodes, which can be problematic if a node is restarted, since the node may no longer contain the same data. We should generate a unique ID for each node and advertise it. Ideally, this ID should be human readable because it will go in the layout.

## Deliverables
- [ ] Generate a unique node ID for each node

### PR DESCRIPTION
**Pull request by no2chem**
## Overview

Description: This PR introduces a node Id field, which is generated on first start and stored in the data store. The node Id is a 128-bit UUID which is represented as a Base64 URL-safe string. The planned use of a base64 string is to append it to a server url, in the form ```<host>:<port>:<node-id>```, which will enable the node id to be checked at connection time during a handshake to be implemented in the future. Currently, the Node Id can be retrieved in using a versionRequest call to a server, and the server can retrieve the node Id from a serverContext.

Why should this be merged: This PR makes it possible to disambiguate servers which have failed, but restarted at the same IP address, eliminating configuration errors from allowing clients to connect to incorrect servers.

Related issue(s) (if applicable): Fixes #1050 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 1]
```diff
@@ -275,12 +271,15 @@ public static void main(String[] args) {
         // Create a common Server Context for all servers to access.
         serverContext = new ServerContext(opts, router);
 
+        // Generate a node ID if necessary.
+        generateNodeId(serverContext);
```
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: Should this be generated here?
Or can we move it down in the ServerContext creation.
This way we will have a similar behaviour between our tests and real servers.

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 2]
```diff
@@ -480,4 +481,21 @@ public static void addManagementServer() {
         managementServer = new ManagementServer(serverContext);
         router.addServer(managementServer);
     }
+
+    /** Generate a Node Id if not present.
+     *
+     * @param context   The server context to use.
+     */
+    private static void generateNodeId(@Nonnull ServerContext context) {
+        String currentId = context.getDataStore().get(String.class, "",
```
Comment (zalokhan) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: Can we move the setter and getter of this NODE_ID to the server context.
Every component will try to consume it.

Comment (no2chem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: I think this is not needed now that there is a helper function.

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java (hunk 1 above)
Comment (no2chem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: Ok - moved it to server context. Going to wait for #1061 to make meaningful tests.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **4ae0bbe**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1059 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/1nxikniF253rXLMtv8ciPEFdoVpv4lpm)

**Pull request merged**